source/AI.cpp

```diff
@@ -2061,95 +2070,192 @@ void AI::KeepStation(Ship &ship, Command &command, const Body &target)
+			if(hasThisAmmo && weapon->SafeRange())
+			{
+				minSafeDistance = max(weapon->SafeRange(), minSafeDistance);
+				splashSpace += outfitSpace;
+			}
 
-			// The missile boat AI should be applied at 1000 pixels range if
-			// all weapons are homing or turrets, and at 2000 if not.
-			double multiplier = (hardpoint.IsHoming() || hardpoint.IsTurret()) ? 1. : .5;
-			shortestRange = min(multiplier * weapon->Range(), shortestRange);
+			// The artillery AI should be applied at 1000 pixels range.
+			// regardless of wether the weapon is homing or not.
+			// The AI works fine with non homing weapons
+			double range = weapon->Range();
+			shortestRange = min(range, shortestRange);
+			if(range > 1000)
+			{
+				shortestArtillery = min(range, shortestArtillery);
+				rangedSpace += outfitSpace;
+			}
+			
+			
 		}
 	}
+	
+	// Calculate this ship's "turning radius"; that is, the smallest circle it
+	// can make while at full speed.
+	double stepsInFullTurn = 360. / ship.TurnRate();
+	double circumference = stepsInFullTurn * ship.Velocity().Length();
+	double diameter = circumference / PI;
+	
 	// If this ship was using the missile boat AI to run away and bombard its
 	// target from a distance, have it stop running once it is out of ammo. This
 	// is not realistic, but it's a whole lot less annoying for the player when
 	// they are trying to hunt down and kill the last missile boat in a fleet.
 	if(isArmed && !hasAmmo)
+	{
 		shortestRange = 0.;
+		shortestArtillery = 0.;
+	}
+	
+	// ArtilleryAI is the AI responsible for handling the behavior of missile boats
+	// and other ships with exceptionally long range weapons such as detainers
+	// The AI shouldn't use artilleryAI if it has no reverse and it's turning
+	// capabilities are very bad. Otherwise it spends most of it's time flying around
+	if(rangedSpace > totalSpace * 0.5 && (ship.MaxReverseVelocity() ||
+			diameter < 0.2 * shortestArtillery))
+		artilleryAI = true;
+	// Don't try to avoid your own splash damage if it means you whould be losing out
+	// on a lot of DPS. Helps with ships with very slow turning and not a lot of splash
+	// weapons being overly afraid of dying.
+	if(minSafeDistance && !(artilleryAI || shortestRange * splashSpace / totalSpace
+			> diameter))
+		minSafeDistance = 0.;
+
+	
+	ship.SetAiAttributes("artilleryAI", artilleryAI);
+	ship.SetAiAttributes("shortestRange", shortestRange);
+	ship.SetAiAttributes("shortestArtillery", shortestArtillery);
+	ship.SetAiAttributes("minSafeDistance", minSafeDistance);
+	ship.SetAiAttributes("turning radius", diameter);
+}
 
+
+
+void AI::Attack(Ship &ship, Command &command, const Ship &target)
+{
+	const Dictionary aiAttributes = ship.AiAttributes();
+	bool artilleryAI = aiAttributes.Get("artilleryAI") == 1;
+	double shortestRange = aiAttributes.Get("shortestRange");
+	double shortestArtillery = aiAttributes.Get("shortestArtillery");
+	double minSafeDistance = aiAttributes.Get("minSafeDistance");
```

Comment:
Why is a dynamic collection used solely to store statically known entries? Use a plain class with fields to store the information; it will be faster and smaller since it won't use any strings.

Reply:
Resolved by https://github.com/HTMLPhoton/endless-sky/commit/48f8af63115a6d22acff07af243adfd984aa22c5